### PR TITLE
Added ColorPropType

### PR DESCRIPTION
**Pull request by shtefanntz**
Problem: https://github.com/facebook/react-native/issues/4708

Solution: Added a ColorPropType that validates the color used by the dev

Notes: 
1) I'm working a Win8.1 machine and couldn't build the react-native using the github repo. As soon as I figure that out, I'll probably figure how to run the tests and how to add some for this feature.
2) It's my first pull request. Be gentle :)

### CODE REVIEW

**Comment by nicklockwood** [comment]
Seems like you and @bestander both had the same idea.

The color validation is a nice idea. I can't decide if it's overkill or not, but I guess since it's only enabled in DEV mode, it's probably not a perf overhead.

**Comment by nicklockwood** [comment]
If you wouldn't mind fixing the lint warnings, I think this will be good to go.

**Comment by bestander** [comment]
There are a few ESLint errors.
Also type number should be allowed for ColorPropType (as in https://github.com/facebook/react-native/issues/4909) because `tinycolor` outputs number and reuses same property.

**Comment by shtefanntz** [comment]
Quick question: I ran eslint and it is clean now, but I can't get 'ColorPropType Required module not found'. Any suggestions? (I running on windows btw)

**Comment by bestander** [comment]
You need to add comment at the header
```
 /**
  * Copyright (c) 2015-present, Facebook, Inc.
  * All rights reserved.
  *
  * This source code is licensed under the BSD-style license found in the
  * LICENSE file in the root directory of this source tree. An additional grant
  * of patent rights can be found in the PATENTS file in the same directory.
  *
  * @providesModule ColorPropType
  */
```

`@providesModule ColorPropType` is the important part

**Comment by shtefanntz** [comment]
@bestander Thanks! This should be added in the 'before you make a pull request' section. I know I read it a few days ago, but I don't remember anything mentioning about it

**Comment by bestander** [comment]
@shtefanntz thanks a lot.
You are right, it should be added (a PR would be welcome ;)

Would you be able to support `number` type for the color checker?
I've had a duplicate PR #4909 that supports a number type.

**Comment by shtefanntz** [comment]
I wasn't sure what you meant when I submitted the last commit. I'll update the PR soon.

**Comment by shtefanntz** [comment]
Btw, I've added the file header, but I still get errors from eslint-bot because of not finding the module :(

**Comment by bestander** [comment]
Yeah, maybe a bug in ESlint bot

**Comment by bestander** [comment]
@vjeux do you reckon ESLint bot could be faulty here?

**Comment by bestander** [comment]
@shtefanntz could you have a look how this checker works in case of array then please?

If I add this check to `Libraries/PullToRefresh/PullToRefreshViewAndroid.android.js`
```
colors: React.PropTypes.arrayOf(ColorPropType),
```
The checker seems to still have the warning.

I reckon you need to access value via `props[propName]`

**Comment by vjeux** [comment]
This is a problem with github way of handling inline comments. Unlike phabricator which clears all the inline comments when you sends an update, github only clears the inline comments on lines that were updated.

In this case, the require didn't change so github left them there. If you have ideas on how we can workaround it, please let me know

**Comment by shtefanntz** [comment]
@bestander Not being able to run the code sure sucks. I updated the signature after looking closer at all the create*Checker functions in ReactPropType.js

**Comment by bestander** [comment]
@shtefanntz awesome!
Thanks.
it_compiles_ship_it

**Comment by bestander** [comment]
You can't run the code?

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit

**Comment by bestander** [comment]
@facebook-github-bot shipit

**Comment by nicklockwood** [comment]
Thanks for the PR. This actually caught a few cases in our internal code where we were using the invalid color name "clear" instead of "transparent". Awesome!

File: Libraries/StyleSheet/ColorPropType.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  */
+'use strict';
+var ReactPropTypes = require('ReactPropTypes');
+var tinycolor = require('tinycolor');
+
+var colorValidator = function (props, propName) {
+  var selectedColor = props[propName];
+  if (selectedColor === null || selectedColor === undefined || selectedColor.toString().trim() === '') {
+    return new Error(
+      `Invalid argument supplied to ${propName}.Expected a string like #123ADF or 'red'.`
```
Comment (vjeux) [inline Libraries/StyleSheet/ColorPropType.js]: nit: space after `.`

File: Libraries/StyleSheet/ColorPropType.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    );
+  }
+
+  if (tinycolor(selectedColor.toString().trim()).isValid()) {
+    return null;
+  }
+
+  return new Error(
+    `Invalid argument supplied to ${propName}.Expected a string like #123ADF or 'red'.`
```
Comment (vjeux) [inline Libraries/StyleSheet/ColorPropType.js]: nit: space after `.`

File: Libraries/StyleSheet/ColorPropType.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  * LICENSE file in the root directory of this source tree. An additional grant
+  * of patent rights can be found in the PATENTS file in the same directory.
+  *
+  * @providesModule ColorPropType
+  */
+'use strict';
+var ReactPropTypes = require('ReactPropTypes');
+var tinycolor = require('tinycolor');
+
+var colorValidator = function (props, propName) {
```
Comment (vjeux) [inline Libraries/StyleSheet/ColorPropType.js]: nit: no space before `(`

File: Libraries/StyleSheet/ColorPropType.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  var selectedColor = props[propName];
+  if (selectedColor === null || selectedColor === undefined || selectedColor.toString().trim() === '') {
```
Comment (vjeux) [inline Libraries/StyleSheet/ColorPropType.js]: Can you check `typeof selectedColor === 'string'` which will remove the need for the first two conditions.

Also, can you do the trim() only once and store it in a variable. This is a pretty hot path so we want to have as few overhead as possible.

Comment (vjeux) [inline Libraries/StyleSheet/ColorPropType.js]: why do you do .toString() here? We should only accept strings, not objects that pretend to be strings

File: Libraries/StyleSheet/ColorPropType.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  if (tinycolor(selectedColor.toString().trim()).isValid()) {
+    return null;
+  }
+
+  return new Error(
+    `Invalid argument supplied to ${propName}.Expected a string like #123ADF or 'red'.`
+  );
+};
+
+var ColorPropType = ReactPropTypes.oneOfType([colorValidator, ReactPropTypes.number]);
```
Comment (vjeux) [inline Libraries/StyleSheet/ColorPropType.js]: We also support arrays apparently: https://github.com/facebook/react-native/blob/0e8b207cc34f230eb2cced3981734b522b601817/Libraries/StyleSheet/processColor.js#L20-L21

File: Libraries/StyleSheet/ColorPropType.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+var colorValidator = function (props, propName) {
+  var selectedColor = props[propName];
+  if (selectedColor === null || selectedColor === undefined || selectedColor.toString().trim() === '') {
+    return new Error(
+      `Invalid argument supplied to ${propName}.Expected a string like #123ADF or 'red'.`
+    );
+  }
+
+  if (tinycolor(selectedColor.toString().trim()).isValid()) {
```
Comment (vjeux) [inline Libraries/StyleSheet/ColorPropType.js]: It seems like you are duplicating a lot of the same logic as in processColor ( https://github.com/facebook/react-native/blob/0e8b207cc34f230eb2cced3981734b522b601817/Libraries/StyleSheet/processColor.js#L17 ). Could you instead make processColor return either the correct color or `null`. Then here you would just check if it's not null.

This way they always stay in sync.

File: Libraries/StyleSheet/ColorPropType.js (hunk 4 above)
Comment (vjeux) [inline Libraries/StyleSheet/ColorPropType.js]: you probably want to invert the condition here and return null. This way you don't have to write the error message twice :)

**Comment by vjeux** [comment]
Sorry for being late to the party reviewing it. I suggested a few changes as right now it doesn't work validate the same inputs as we currently support on colors.

**Pull request closed without merging**